src/config/command.zig

```diff
@@ -0,0 +1,291 @@
+    pub fn string(
+        self: *const Self,
+        alloc: Allocator,
+    ) Allocator.Error![:0]const u8 {
+        return switch (self.*) {
+            .shell => |v| try alloc.dupeZ(u8, v),
+            .direct => |v| try std.mem.joinZ(alloc, " ", v),
+        };
+    }
+
+    /// Get an iterator over the arguments array. This may allocate
+    /// depending on the active tag of the command.
+    ///
+    /// For direct commands, this is very cheap and just iterates over
+    /// the array. There is no allocation.
+    ///
+    /// For shell commands, this will use Zig's ArgIteratorGeneral as
+    /// a best effort shell string parser. This is not guaranteed to be
+    /// 100% accurate, but it works for common cases. This requires allocation.
+    pub fn argIterator(
+        self: *const Self,
+        alloc: Allocator,
+    ) Allocator.Error!ArgIterator {
+        return switch (self.*) {
+            .direct => |v| .{ .direct = .{ .args = v } },
+            .shell => |v| .{ .shell = try .init(alloc, v) },
+        };
+    }
+
+    /// Iterates over each argument in the command.
+    pub const ArgIterator = union(enum) {
+        shell: std.process.ArgIteratorGeneral(.{}),
+        direct: struct {
+            i: usize = 0,
+            args: []const [:0]const u8,
+        },
+
+        /// Return the next argument. This may or may not be a copy
+        /// depending on the active tag. If you want to ensure that every
+        /// argument is a copy, use the `clone` method first.
+        pub fn next(self: *ArgIterator) ?[:0]const u8 {
+            return switch (self.*) {
+                .shell => |*v| v.next(),
+                .direct => |*v| {
+                    if (v.i >= v.args.len) return null;
+                    defer v.i += 1;
+                    return v.args[v.i];
+                },
+            };
+        }
+
+        pub fn deinit(self: *ArgIterator) void {
+            switch (self.*) {
+                .shell => |*v| v.deinit(),
+                .direct => {},
+            }
+        }
+    };
+
+    pub fn clone(
+        self: *const Self,
+        alloc: Allocator,
+    ) Allocator.Error!Self {
+        return switch (self.*) {
+            .shell => |v| .{ .shell = try alloc.dupeZ(u8, v) },
+            .direct => |v| direct: {
+                const copy = try alloc.alloc([:0]const u8, v.len);
+                for (v, 0..) |arg, i| copy[i] = try alloc.dupeZ(u8, arg);
+                break :direct .{ .direct = copy };
+            },
+        };
+    }
+
+    pub fn formatEntry(self: Self, formatter: anytype) !void {
+        switch (self) {
+            .shell => |v| try formatter.formatEntry([]const u8, v),
+
+            .direct => |v| {
+                var buf: [4096]u8 = undefined;
+                var fbs = std.io.fixedBufferStream(&buf);
+                const writer = fbs.writer();
+                writer.writeAll("!") catch return error.OutOfMemory;
```

Comment:
`direct:` here?

Reply:
Good catch, also added unit tests for this.